packages/paima-sdk/paima-events/src/app-events.ts

```diff
@@ -0,0 +1,200 @@
+}
+
+export function toTopicHash<T extends LogEvent<LogEventFields<TSchema>[]>>(event: T): string {
+  return keccak_256(toSignature(event));
+}
+
+// this generates the type for the events import.
+export const generateAppEvents = <
+  const T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+>(
+  entries: T
+): {
+  [K in T[number] as K['name']]: (ReturnType<typeof toPath<K, typeof TopicPrefix.App>> & {
+    definition: T[0];
+    topicHash: string;
+  })[];
+} => {
+  let result: Record<string, any> = {}; // we can't know the type here
+  for (const event of entries) {
+    if (!result[event.name]) {
+      result[event.name] = [];
+    }
+
+    result[event.name].push({
+      ...toPath(TopicPrefix.App, event),
+      // keep the original definition around since it's nicer to work with, it
+      // also has the advantage that it allows recovering the initial order in
+      // case the signature/topicHash needs to be computed again, which can't be
+      // done from the path (since you don't know which non indexed fields go in
+      // between each indexed field).
+      definition: event,
+      // we add this to avoid having to re-compute it all the time.
+      topicHash: toTopicHash(event),
+    });
+  }
+  return result as any;
+};
+
+// create payload for the stf from an object.
+export function createEventForStf<T extends ReturnType<typeof genEvent>>(
+  address: `0x${string}`,
+  event: T,
+  fields: KeypairToObj<T['fields']>
+): {
+  address: `0x${string}`;
+  data: { name: string; fields: KeypairToObj<T['fields']>; topic: string };
+} {
+  return { address, data: { name: event.name, fields, topic: toTopicHash(event) } };
+}
+
+// sum type for all the overloads of a fixed event name
+type UnionForOverloadedEvents<
+  T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+  N extends string,
+> = T extends (infer U)[]
+  ? U extends { name: N; fields: infer V }
+    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      V extends { name: string; type: any }[]
+      ? KeypairToObj<V>
+      : never
+    : never
+  : never;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type KeypairToObj<T extends { name: string; type: any }[]> = {
+  [K in T[number] as K['name']]: Static<K['type']>;
+};
+
+type EmitFunction<
+  T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>,
+  Name extends string,
+> = {
+  emit: (name: Name, address: `0x${string}`, fields: UnionForOverloadedEvents<T, Name>) => void;
+};
+
+type EventQueue<T extends ReadonlyArray<LogEvent<LogEventFields<TSchema>[]>>> = EmitFunction<
+  T,
+  T[number]['name']
+> & {
+  get: () => {
+    address: `0x${string}`;
+    data: { name: string; fields: KeypairToObj<T[number]['fields']>; topic: string };
+  }[];
+};
+
+// Typed stateful builder that handles overloaded topics. There is a performance
+// cost of using this, since it dynamically finds the matching overloaded topic
```

Comment:
I've been thinking that I included the topic because it was part of the spec/issue. But to be honest I'm not sure how much sense it makes. Maybe moving this logic to the engine would be better? And then this wouldn't be a concern at all. Because currently the user can submit an event with the wrong topic and there is no validation for it at the engine level (there is the validation that the topic exists for that name, but not that the actual fields match).